src/frontends/lean/builtin_cmds.cpp

```diff
@@ -393,10 +393,12 @@ static environment eval_cmd(parser & p) {
     out.report();
     if (io_result_is_error(r.raw())) {
         message_builder msg = p.mk_message(p.cmd_pos(), p.pos(), ERROR);
-        object * err = lean_io_error_to_string(io_result_get_error(r.raw()));
-        msg << string_to_std(err);
+        object * err = io_result_get_error(r.raw());
+        inc_ref(err);
+        object * str = lean_io_error_to_string(err);
+        msg << string_to_std(str);
         msg.report();
-        dec_ref(err);
+        dec_ref(str);
```

Comment:
You can drop the `inc`+`dec`. `lean_io_error_to_string` consumes the reference.

Reply:
`inc_ref(err)` is exactly because `lean_io_error_to_string` it consumes the argument. Meanwhile, `r` still indirectly points to it and its destructor will eventually try to deallocate it. And `dec_ref(str)` is because `lean_io_error_to_string` produces a new string which will not be deallocated when `r` goes out of scope.

I think this was the source of my confusion too. We're dealing with three objects: the result of the IO computation, the error it contains and the string rendering of that error. There's an ownership relation between the IO result and the error (which could cause a double free) but none between the error and the string (and if there was one, I believe `lean_io_error_to_string` should increase its reference count before returning so we should decrease it anyway).

Reply:
Oh, yeah, I misread